elastic/kibana
# [Runtime field editor] Composite runtime in Kibana Data Views

This PR adds support in Kibana Data views (former "Index pattern") for composite runtime fields.

It is the implementation of the https://github.com/elastic/kibana/discussions/99462 discussion.

## The problem

The new "composite" runtime fields allows a user to [emit multiple values from a script](https://github.com/elastic/elasticsearch/pull/75108). This means that when creating one runtime field of `type: "composite"` what the user is actually doing is creating one **or multiple** runtime fields behind a namespace (the name of the runtime field).


When we attach a composite runtime field to a search request:

```js
GET my-index-000001/_search
{
  "runtime_mappings": {
    "myComposite" : {
        "type" : "composite",
        "script" : "emit('day_of_week', 'someValue'); emit('anotherField',  1234);",
        "fields" : {
            "day_of_week" : {
                "type" : "keyword"
            },
            "anotherField" : {
                "type" : "long"
            }
        }
      }
  },
  "aggs": {
    "day_of_week": {
      "terms": {
        "field": "myComposite.day_of_week"
      }
    }
  }
}
```

we are creating **two runtime fields** for the user: `myComposite.day_of_week` and `myComposite.anotherField`. The former of type `keyword` and the second of type `long`.

We need a way to store in the Kibana Data view both the composite runtime field (**which is not a field** but a holder of fields) and the fields themselves. 

## How to test

* Navigate to Discover
* Create a new field on the index pattern
* Give a name (e.g. `"log"`) and set the type to `composite`
* Toggle the "Set value" and add a script

```
emit('field1', 'hello');
emit('field2', 'world');
```

<img width="1081" alt="Screenshot 2021-08-26 at 15 28 41" src="https://user-images.githubusercontent.com/2854616/130982000-3f687975-5404-43f2-ade0-fb1d025c0f61.png">


* Save the runtime field
* The 2 fields should appear in the list of fields (under `"log.field1"` and `"log.field2"`
* You should be able to select them as any other field

<img width="1337" alt="Screenshot 2021-08-26 at 15 29 22" src="https://user-images.githubusercontent.com/2854616/130981934-b29e1aa3-2def-453e-8261-6b3698df4810.png">


* The 2 fields should be available in the Search bar

<img width="410" alt="Screenshot 2021-08-26 at 15 29 01" src="https://user-images.githubusercontent.com/2854616/130981884-75d0bf93-4010-456a-9a22-af2f5458c7b1.png">


* You should **not** be able to edit or delete them (work for future PR)

## Future work

This PR is going to a feature branch. Future PR will add

* [ ] Component integration tests
* [ ] Server CRUD routes to Get/Create/Update/Delete composite runtime in the Data view
* [ ] UI to set the format, type, custom label and popularity to sub fields
* [ ] UI (modals) to indicate how to delete/edit a subfield

Fixes https://github.com/elastic/kibana/issues/99177
